### PR TITLE
Provide a better error message if a DSLError is raised

### LINKED ISSUE #6457: Using specific git repository with Podspec that has syntax error will say no Podspec found
# Report

## What did you do?

Trying to install a pod from a specific git repository, that stores the Podspec file. This is the entry in Podfile:

```
pod 'Kvitto', :git => 'https://github.com/cocoanetics/kvitto.git'
```

After running `pod install`, this is the output:
```
Pre-downloading: `Kvitto` from `https://github.com/cocoanetics/kvitto.git`
[!] Unable to find a specification for 'Kvitto'.
```

## What did you expect to happen?

The specification is available on the URL, but the Podspec file has syntax error. The error output by Cocoapods, should clearly state that Podspec was found, but was unable to be parsed.

## What happened instead?
Cocoapods error was to unable to find specification. Since Podspec file is there, I was looking for error in wrong places for quite a while.

### Stack

```
   CocoaPods : 1.2.0
        Ruby : ruby 2.0.0p648 (2015-12-16 revision 53162) [universal.x86_64-darwin16]
    RubyGems : 2.0.14.1
        Host : Mac OS X 10.12.3 (16D32)
       Xcode : 8.2.1 (8C1002)
         Git : git version 2.10.1 (Apple Git-78)
Ruby lib dir : /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib
Repositories : 
               master - https://github.com/CocoaPods/Specs.git @ a0eeb5b2f70ec81c193486cf869f6ca29ed38293
```

### Installation Source

```
Executable Path: /usr/local/bin/pod
```

### Plugins

```
cocoapods-deintegrate : 1.0.1
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.0.0
cocoapods-trunk       : 1.1.2
cocoapods-try         : 1.1.0
```

### Podfile

```ruby
platform :ios, '9.0'

inhibit_all_warnings!
use_frameworks!

source 'https://github.com/CocoaPods/Specs.git'

abstract_target 'Application' do
    pod 'Kvitto', :git => 'https://github.com/legoless/Kvitto.git'

    target 'FirstBlubs'

end
```

### PR DESCRIPTION
**Pull request by dnkoutso**
closes #6457

### CODE REVIEW

File: spec/unit/external_sources/abstract_external_source_spec.rb [hunk 1]
```diff
@@ -26,6 +26,22 @@ module Pod
...
+          e = @subject.send(:pre_download, config.sandbox)
+          e.message.should.include "Failed to load 'Reachability' podspec:"
+          e.message.should.include "Invalid `Reachability.podspec` file:"
+        }
+      end
+
+      it 'raises a generic error if a podspec was not found' do
+        Downloader.stubs(:download).raises(Pod::Downloader::DownloaderError.new('Some generic exception'))
+        should.raise(Informative) {
+          @subject.send(:pre_download, config.sandbox).message.should == "Uable to find a specification for 'Reachability'."
```
Comment (dnkoutso) [inline spec/unit/external_sources/abstract_external_source_spec.rb]: @benasher44 not sure what I am doing wrong here but whatever I compare the message too the test is always green.

File: lib/cocoapods/sandbox/podspec_finder.rb [hunk 2]
```diff
@@ -12,13 +12,9 @@ def podspecs
         @specs_by_name = {}
         spec_files = Pathname.glob(root + '{,*}.podspec{,.json}')
         spec_files.sort_by { |p| -p.to_path.split(File::SEPARATOR).size }.each do |file|
-          begin
-            spec = Specification.from_file(file)
-            spec.validate_cocoapods_version
-            @specs_by_name[spec.name] = spec
-          rescue => e
-            UI.warn "Unable to load a podspec from `#{file.basename}`, skipping:\n\n#{e}"
```
Comment (dnkoutso) [inline lib/cocoapods/sandbox/podspec_finder.rb]: This was unnecessarily swallowing all exceptions making unable for callers to handle errors properly or differently.

File: lib/cocoapods/external_sources/abstract_external_source.rb [hunk 3]
```diff
@@ -111,11 +111,15 @@ def pre_download(sandbox)
         title = "Pre-downloading: `#{name}` #{description}"
         UI.titled_section(title,  :verbose_prefix => '-> ') do
           target = sandbox.pod_dir(name)
-          download_result = Downloader.download(download_request, target, :can_cache => can_cache)
+          begin
+            download_result = Downloader.download(download_request, target, :can_cache => can_cache)
+          rescue Pod::DSLError => e
+            raise Informative, "Failed to load '#{name}' podspec: #{e.message}"
```
Comment (dnkoutso) [inline lib/cocoapods/external_sources/abstract_external_source.rb]: using `#{e}` here prints:

```
[!] Failed to load 'Kvitto' podspec: Pod::DSLError
```

Which is not helpful at all.

File: lib/cocoapods/sandbox/podspec_finder.rb (hunk 2 above)
Comment (dnkoutso) [inline lib/cocoapods/sandbox/podspec_finder.rb]: Well the more I think about this maybe I can keep this to maintain the prior semantics that `@specs_by_name` can include a `nil` spec but perhaps raise only if `e` is type of `DSLError`

File: lib/cocoapods/external_sources/abstract_external_source.rb (hunk 3 above)
Comment (benasher44) [inline lib/cocoapods/external_sources/abstract_external_source.rb]: Can you add this `begin` `rescue` to `store_podspec` in this file too? That'll handle similar cases for `PathSource` and `PodspecSource`.

File: lib/cocoapods/sandbox/podspec_finder.rb (hunk 2 above)
Comment (benasher44) [inline lib/cocoapods/sandbox/podspec_finder.rb]: I think removing this is fine. This only appears to be called from `cocoapods/downloader.rb`.

Comment (dnkoutso) [inline lib/cocoapods/sandbox/podspec_finder.rb]: 👍

File: lib/cocoapods/external_sources/abstract_external_source.rb (hunk 3 above)
Comment (dnkoutso) [inline lib/cocoapods/external_sources/abstract_external_source.rb]: Will take care of it.

Comment (dnkoutso) [inline lib/cocoapods/external_sources/abstract_external_source.rb]: @benasher44 do you mean wrap the `sandbox.store_podspec(name, spec.to_pretty_json, true, true)` call with `begin` and `rescue` or do you mean line 123 `store_podspec(sandbox, spec)` to be wrapped?

Comment (dnkoutso) [inline lib/cocoapods/external_sources/abstract_external_source.rb]: hm I guess you mean both :)

**Comment by dnkoutso** [comment]
@benasher44 updated.

**Comment by dnkoutso** [comment]
@DanToml @benasher44 any objections on merging this?

**Comment by benasher44** [comment]
Merge!

**Pull request merged**